compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -691,11 +691,65 @@ public boolean isArrayTypesAssignable(BType source, BType target, List<TypePair>
 
     private boolean isFunctionTypeAssignable(BInvokableType source, BInvokableType target,
                                              List<TypePair> unresolvedTypes) {
+        // For invokable types with typeParam parameters, we have to check whether the source param types are
+        // covariant with the target param types.
+        if (containsTypeParams(target)) {
+            // TODO: 7/4/19 See if the below code can be generalized to avoid code duplication
+            if (source.paramTypes.size() != target.paramTypes.size()) {
+                return false;
+            }
+
+            for (int i = 0; i < source.paramTypes.size(); i++) {
+                BType sourceParam = source.paramTypes.get(i);
+                BType targetParam = target.paramTypes.get(i);
+                boolean isTypeParam = Symbols.isFlagOn(targetParam.tsymbol.flags, Flags.TYPE_PARAM);
```

Comment:
Shall we use TypeParamAnalyzer.isTypeParam() logic here? it is a NPE safe operation.